### ISSUE TITLE
Allow ShouldThrow and ShouldNot throw for Func<t> as well as action

### ISSUE THREAD

**Issue by SeanFarrow** [dennisdoomen/FluentAssertions#422]
Hi,

It would be nice to have the ShouldThrow and ShouldNotThrow for functions as well as actions and Funct<Task.
Cheers
sEan.

**Comment by somewhatabstract** [comment]
How would that work? What would supply the parameters for the function?

Can you provide an example of how you envision this being used?

**Comment by somewhatabstract** [comment]
I see that #421 has a similar request and for `Func<TReturn>`, I can see how this might work, though I am unsure of why those situations could not just be rewritten in `Action` form.

**Comment by SeanFarrow** [comment]
How would you rewrite something like that in action form, when the function returns.
Cheers
Sean.

**Comment by somewhatabstract** [comment]
If you just write:

```
Action underTest = () => ThingThatReturns();
```

the return value is ignored.

**Comment by eyalsk** [comment]
@somewhatabstract Yeah, for a function try that with a property or an indexer and your example will fail.

**Comment by jnyrup** [comment]
For the most parts, we should be able to reuse the function signatures from the actions assertions, but how about `ShouldNotThrow` for funcs.

What should the function signature be?

So far my best guess is something like this:
```csharp
public AndWhichConstraint<FuncAssertions<TResult>, TResult> ShouldNotThrow(string because = "", params object[] becauseArgs)

Func<T> ctor = () => new MyCollection();
ctor.ShouldNotThrow()
.Which.Should().BeEmpty();
```
